keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1793 @@
+type PodSpec struct {
+  Containers []Container `json:...`
+}
+
+type Container struct {
+  // ...
+
+  Ports []ContainerPort `json:...`
+
+  // ...
+}
+
+type ContainerPort struct {
+  // ...
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  HostPort int32 `json:...`
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  ContainerPort int32 `json:...`
+
+  // ...
+}
+```
+
+In this example, both `+validation`, `+minimum` and `+maximum` are IDL tags.
+
+The declarative validation rules will be used by the kube-apiserver to
+validate API requests.
+
+The declarative validation rules will also be included in the published OpenAPI:
+
+```json
+// Pod Spec
+"openAPIV3Schema": {
+  "type": "object",
+  "x-kubernetes-validations": [
+    {
+      "rule": "!self.hostNetwork || self.containers.all(c, c.containerPort.all(cp, cp.hostPort == cp.containerPort))"
+    }
+  ],
+}
+
+...
+
+// Container Port
+"openAPIV3Schema": {
+  "type": "object",
+  "properties": {
+    "hostPort": {
+      "minimum": 1,
+      "maximum": 65535
+      ...
+    }
+    "containerPort": {
+      "minimum": 1,
+      "maximum": 65535
+      ...
+    }
+  }
+}
+```
+
+It is important to note that the declarative validation will use the same set of
+validation options available today for CRDs. namely:
+
+- [JSON Schema value validations](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00) (e.g. `format`, `maxItems`)
+- [CEL Validation Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) (e.g. `self.replicas <= self.maxReplicas`)
+- Relevant [merge strategy declarations](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) (e.g. `listType=set`)
+
+## Motivation
+
+Kubernetes API validation rules are currently written by hand, which makes them
+difficult for users to access directly.
+
+Declarative validation will benefit Kubernetes maintainers:
+
+- It will make it easier to develop, maintain and review APIs.
+- It will make it easier programatically inspect and analyze the API, enabling
+  new tools and improved documentation.
+- It will enable improvements to the API machinery. For example, a feature like
+  ratcheting validation will become more tractable to implement because the
+  feature can be implemented once in the declarative validation subsystem
+  rather than piecemeal across the 15k lines of hand written validation code.
+
+Declarative validation will also benefit Kubernetes users:
+
+- It will give users direct access to the actual API validation rules, which
+  are currently only available to developers willing and able to find and read
+  the hand written validation rules.
+- It will enable clients to perform validation of native types earlier in
+  development worflows ("shift-left"), such as at with a Git pre-submit linter.
+- It will improve API composition. In particular CRDs that embed native types
```

Comment:
For the case of validation, we have the fallback of using CEL rules at point-of-use to refer to fields within the template. 

For defaults, we do need to solve this problem. I'm pondering ways this could be done but very open to feedback on that front, assuming it is not feasible to separate the types.

Reply:
I wonder if something like:

```
type PodSpec struct {
      TerminationGracePeriodSeconds int32;
}

type Pod struct {
    // +default(self.TerminationGracePeriodSeconds)=30
    Spec PodSpec
}
```

I don't know how that would turn into docs.  

Or maybe we want something more explicit:

```
type PodSpec struct {
    // +default=30
    TerminationGracePeriodSeconds int32;
}

type Pod struct {
    // inherits all defaults 
    Spec PodSpec
}

type Pod struct {
    // do not inerit defaults
    // +default={}
    Spec PodSpec `k8s:"no-defaults"`
}
```

Reply:
As far as I can tell, that should already work. I feel like it's a bug if it doesn't. We should probably add a test for that.

Reply:
We should put the defaults discussion in a different place and think-thru an example of how to do contextual validation here.

Reply:
Also, since we have historical baggage, we may not be able to do "obvious" things like I suggested above.  I recently changed ONE field to stop defaulting in *.PodSpec and instead only in Pod.PodSpec, and I broke things (doing a diff of old saved state with new dry-run results - the field was marked absent).